drivers/ethernet/eth_renesas_ra.c

```diff
@@ -0,0 +1,370 @@
+		LOG_ERR("Failed to init ether - R_ETHER_CallbackSet fail");
+	}
+}
+
+static int renesas_ra_eth_tx(const struct device *dev, struct net_pkt *pkt)
+{
+	fsp_err_t err = FSP_SUCCESS;
+	struct renesas_ra_eth_context *ctx = dev->data;
+	uint16_t len = net_pkt_get_len(pkt);
+	static uint8_t tx_buf[NET_ETH_MAX_FRAME_SIZE] __aligned(4);
+
+	if (net_pkt_read(pkt, tx_buf, len)) {
+		goto error;
+	}
+
+	/* Check if packet length is less than minimum Ethernet frame size */
+	if (len < NET_ETH_MINIMAL_FRAME_SIZE) {
+		/* Add padding to meet the minimum frame size */
+		memset(tx_buf + len, 0, NET_ETH_MINIMAL_FRAME_SIZE - len);
+		len = NET_ETH_MINIMAL_FRAME_SIZE;
+	}
+
+	if (ctx->link_up == true) {
+		err = R_ETHER_Write(&ctx->ctrl, tx_buf, len);
+		if (err != FSP_SUCCESS) {
+			goto error;
+		}
+	} else {
+		return -1;
+	}
+
+	return 0;
+
+error:
+	LOG_ERR("Writing to FIFO failed");
+	return -1;
+}
+
+static const struct ethernet_api api_funcs = {
+	.iface_api.init = renesas_ra_eth_initialize,
+	.get_capabilities = renesas_ra_eth_get_capabilities,
+	.send = renesas_ra_eth_tx,
+};
+
+static void renesas_ra_eth_isr(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	ether_eint_isr();
+}
+
+static struct net_pkt *renesas_ra_eth_rx(const struct device *dev)
+{
+	fsp_err_t err = FSP_SUCCESS;
+	struct renesas_ra_eth_context *ctx;
+	struct net_pkt *pkt = NULL;
+	uint32_t len = 0;
+	static uint8_t rx_buf[NET_ETH_MAX_FRAME_SIZE] __aligned(4);
+
+	__ASSERT_NO_MSG(dev != NULL);
+	ctx = dev->data;
+	__ASSERT_NO_MSG(ctx != NULL);
+
+	err = R_ETHER_Read(&ctx->ctrl, rx_buf, &len);
+	if ((err != FSP_SUCCESS) && (err != FSP_ERR_ETHER_ERROR_NO_DATA)) {
+		LOG_ERR("Failed to read packets");
+		goto out;
+	}
+
+	pkt = net_pkt_rx_alloc_with_buffer(ctx->iface, len, AF_UNSPEC, 0, K_MSEC(100));
+	if (!pkt) {
+		LOG_ERR("Failed to obtain RX buffer");
+		goto out;
+	}
+
+	if (net_pkt_write(pkt, rx_buf, len)) {
+		LOG_ERR("Failed to append RX buffer to context buffer");
+		net_pkt_unref(pkt);
+		pkt = NULL;
+		goto out;
+	}
+
+out:
+	if (!pkt) {
+		eth_stats_update_errors_rx(ctx->iface);
+	}
+
+	return pkt;
+}
+
+static void renesas_ra_eth_thread(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	const struct device *dev = p1;
+	struct net_if *iface;
+	int res;
+	struct net_pkt *pkt = NULL;
+	struct renesas_ra_eth_context *ctx = dev->data;
+
+	while (true) {
+		res = k_sem_take(&ctx->rx_sem, K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
+		if (res == 0) {
+			/* semaphore taken, update link status and receive packets */
+			if (ctx->link_up != true) {
+				res = R_ETHER_LinkProcess(&ctx->ctrl);
+				if (res == FSP_SUCCESS) {
+					ctx->link_up = true;
+					net_eth_carrier_on(ctx->iface);
+				}
+			}
+
+			pkt = renesas_ra_eth_rx(dev);
+
+			if (pkt != NULL) {
+				iface = net_pkt_iface(pkt);
+				res = net_recv_data(iface, pkt);
+				if (res < 0) {
+					net_pkt_unref(pkt);
+				}
+			}
+		} else if (res == -EAGAIN) {
+			res = R_ETHER_LinkProcess(&ctx->ctrl);
```

Comment:
should be using zephyr phy API in the driver, no new vendor hal phy apis please, users hate this

Reply:
@decsny , Thank you for you feedback, we change to implement mdio driver and now is using phy API from Zephyr, could you please help to recheck